### ISSUE TITLE
Not finding subdirectories in module

### ISSUE THREAD

**Issue by aleross** [claudio-silva/grunt-angular-builder#16]
Hi, first I want to say this is a fantastic tool! Exactly what we were looking for.

My directory structure looks like this:

```src
 |--moduleDirectory
 |  |--servicesDirectory
 |  |    service1.js
 |  |    services2_and_3.js
 |  |--directivesDirectory
 |  |    my_directives.js
 |  |    some_other_directive.js
 |  module.js
```

Where I have many moduleDirectories, and each one has a module.js. Each of my modules is assigned to a variable `var myModule = angular.module()` which my services etc are chained on `myModule.service()`. For some reason, angular-builder is correctly finding the top-level module but not the subdirectories.

Any tips? Much appreciated.

**Comment by claudio-silva** [comment]
Hi!
I'm glad you like the project.

A side note: you could surround your directory tree on the post above with triple backticks, so that it is rendered with a fixed-width font. The way it is formatted now, it is difficult to be sure about the relative file locations.

I guess the `module.js` file is inside the `moduleDirectory`, which has 2 subdirs, each one containing multiple js files, all defining services and directives for the module above.
This should work just fine, and I have insufficient information from you (i.e. source code to look at) to guess what may be  happening here, but I can make a few guesses:

1. your Grunt configuration may define glob patterns that prevent Grunt from finding all your javascript files.
2. The module name doesn't match between `module.js` and each of your service files.
3. Your main application is not "requiring" all your modules (directly or indirectly) so, although they are being found by Angular Builder, they are also being discarded by the "tree shaking" algorithm.
4. it may be caused by the way you declare/reference your modules.

Suggestions:

1. try running Grunt in verbose mode to see exactly what files are being processed and get a better glimpse of what's happening "under the hood". See [Debugging build failures](../#debugging-build-failures) on the Readme.
2. write your modules with the recommended syntax and structure, as it is explained in [Writing build enabled code](../wiki/Writing-build-enabled-code) on the Wiki.
3. this doesn't mean you can't write your code the way you are writing it. Without seeing it I can only guess it should work, but please pay special attention to [this section of the documentation](../wiki/Writing-build-enabled-code#assigning-module-references-to-variables).

If all of this fails, I suggest you add a test case to the project, with the minimal set of files and source code to reproduce the problem, and post a link here to your commit/repo, so that I can properly test it.

Regards,
Cláudio

**Comment by aleross** [comment]
Thanks for your well-structured and quick response Claudio. I updated my post with the triple backticks.

I ran Grunt in debug mode and can see that angular-builder is ignoring the files I'd like included, so I'm guessing it has to do with my module syntax and structure. This is what it looks like, given a pseduo module with a unique name `Endpoint` and the service is a unique service name `EndpointCall`

**modules/Endpoint/Endpoint.js:**

`var endpoint = angular.module('Endpoint', []);`

**modules/Endpoint/services/EndpointService.js**:

`endpoint.service('EndpointService', [], function () {});`


Then, the verbose grunt task output looks like:

```
Reading modules/Endpoint/Endpoint.js...OK
Reading modules/Endpoint/services/EndpointService.js...OK
Ignored file: modules/Endpoint/services/EndpointService.js
Including module Endpoint.
Scanning Endpoint for non-angular script dependencies...
Validating modules/Endpoint/Endpoint.js...OK
```

I read through the angular-builder documentation a few times and it seems like storing the module in a variable should work, so I'm not sure why it might not be picking it up. Appreciate any input.

**Comment by aleross** [comment]
If nothing simple looks off to you then I'll create a small set of files to try to reproduce.

**Comment by claudio-silva** [comment]
Thank you for your helpful explanation. I can see the problem now.
Although you may store a reference to your module in a variable, that variable is confined to the file where it is declared. You may not assign your module to a global variable (nor is it a good practice).
Furthermore, Angular Builder has specific code to detect the declaration of global variables and forbid it (only on 'release' mode), as your application code would behave differently on release and debug builds, due to the way the builder wraps and optimizes code on release builds.
Note: although you can still define global variables by specifically storing them on the `window` object, the builder will not make any use of them.
The recommended way to structure your code is explained in the wiki. Please read [Writing build enabled code](../wiki/Writing-build-enabled-code). Although you may deviate from it, you should consider using it, if possible, as it reflects good programming practices (IMHO) in the context of Angular.js application architecture.
So, in summary, to effectively use Angular Builder, you should always have, at least, one `angular.module('name')` reference to a module on each file related to it.

**Comment by aleross** [comment]
That fixed it, thanks Claudio. I had no idea you could reuse `angular.module()` in service etc declarations, I must have missed that day in angular 101 :). I'll be taking out those nasty global vars.

**Comment by claudio-silva** [comment]
Great! I'm glad I could help.
I've updated the documentation to explain this, as it is quite possible that other people may also face this same problem.
See [this section](../wiki/Writing-build-enabled-code#assigning-module-references-to-variables) and the following one.

**Comment by claudio-silva** [comment]
(I noticed the link may not scroll the page to the relevant section unless you refresh the browser)
